Proposal load/store masked

DiamonDinoia opened the pull request:
Dear @serge-sans-paille,

I made a rough implementation of masked load/store. Before I hash it out. Can I have some early feedback? 

Thanks,
Marco

serge-sans-paille commented:
interesting. Before going into the details, what are the memory effects of a masked load in terms of read memory and value stored? Are the masked elements set to 0 ? to an undefined value ? AVX etc seems to set the value to zero?

I'm not sold on the common implementation which looks quite heavy in scalar operation. I can see that we can't do a plain load followed by an and because it could lead to access to unallocated memory. If the mask were constant, we could optimize statically some common patterns, but with a dynamic mask as you propose...

DiamonDinoia commented:
Some thoughts:
1. Undefined for masked values. Since depending on the operations 0 or 1 might be the correct values. In that case they could use the mask itself to initialize the values.
2. We could remove the dynamic mask entirely. I added for completeness.
3. We could do a la vcl and have a load partial, store partial where we just optimize for head and tail. I preferred this solution as I'm assuming xsimd users know the performance implications of the API.
4. I would use undefined for the values (I might change the api accordingly) because emaing I want to polulate the even elements from one memory location and the odd from another. Masked loads (I think) are faster than a gather.
5. For now this is fast only on avx, av2 but for sse even if it heavy on scalar it is slow only when reading bytes or short. We could optimize these cases.
6. I'm not sure about sve/neon

DiamonDinoia commented:
I don't think the error is due to anything related to this PR. But, I like the looks of it now. Ready to review!

serge-sans-paille commented:
@DiamonDinoia we plan to do a release in the forthcoming weeks. Do you want this PR to be part of it?

JohanMabille commented:
> I would like to have a release by mid-November

That is more or less what we planned ;)

serge-sans-paille commented on include/xsimd/arch/xsimd_avx2.hpp:
```diff
@@ -116,6 +117,160 @@ namespace xsimd
+                return load<A>(mem, Mode {});
+            }
+            // confined to lower 128-bit half (4 lanes) → forward to SSE
+            else XSIMD_IF_CONSTEXPR(mask.countl_zero() >= 4)
```
Thinking of it, we could remove the need for `countl_zero` by just masking, something like

```
if((mask.mask() & 0xF0) == 0)
```

which would be a decent implementation for
```
if(mask.lower_half() == 0) // or lower_mask ?
```
which has the nice property of not encoding any "magical value"

serge-sans-paille commented on include/xsimd/arch/xsimd_avx.hpp:
```diff
@@ -865,6 +881,134 @@ namespace xsimd
+            else XSIMD_IF_CONSTEXPR(mask.countr_zero() >= 2)
+            {
+                constexpr auto mhi = ::xsimd::detail::upper_half<sse2>(mask);
+                const batch<double, sse2> hi(_mm256_extractf128_pd(src, 1));
```
that's a upper_half, right?

serge-sans-paille commented on include/xsimd/arch/xsimd_avx.hpp:
```diff
@@ -865,6 +881,134 @@ namespace xsimd
+            else XSIMD_IF_CONSTEXPR(mask.countl_zero() >= 2)
+            {
+                constexpr auto mlo = ::xsimd::detail::lower_half<sse2>(mask);
+                const batch<double, sse2> lo(_mm256_castpd256_pd128(src));
```
that's a lower_half, right?

serge-sans-paille reviewed (commented):
A new wave of comments. I have the feeling that we could simplify the implementation a lot by making it more generic.

DiamonDinoia commented on include/xsimd/arch/xsimd_avx.hpp:
```diff
@@ -865,6 +881,134 @@ namespace xsimd
+            {
+                _mm256_maskstore_pd(mem, mask.as_batch(), src);
+            }
+        }
```
I'll also have to think if the size of the element matters. I'll have a look next week

serge-sans-paille commented on include/xsimd/arch/xsimd_avx.hpp:
```diff
@@ -865,6 +881,134 @@ namespace xsimd
+            {
+                constexpr auto mhi = ::xsimd::detail::upper_half<sse4_2>(mask);
+                const auto hi = load_masked(mem + 4, mhi, convert<float> {}, Mode {}, sse4_2 {});
+                return batch<float, A>(detail::merge_sse(batch<float, sse4_2>(0.f), hi));
```
I don't know if we could use a similar trick to `_mm256_castps128_ps256 ` here

DiamonDinoia commented:
> @AntoinePrv once you update that branch we can do a release \o/

I will work on this asap. I just have something to deliver on another project this week so it might take a bit for me to get back to this.

DiamonDinoia committed:
Feat: Load/Store masked API
1. Adds new masked API compile time masks (store_masked and load_masked)
2. General use case optimization
3. New tests
4. x86 kernels
5. Adds new APIs to batch_bool_constant for convenience resembling #include<bit>
6. Tests the new APIs

DiamonDinoia commented:
I removed most of the duplication at least I stopped when I thought things were getting out of hand. I think in the future we might offer direct access to masked load with dynamic mask. Which is basically exposing the internal helper functions but adapting them to the xsimd convention.

DiamonDinoia commented on include/xsimd/arch/xsimd_avx512f.hpp:
```diff
@@ -872,11 +1010,11 @@ namespace xsimd
             XSIMD_IF_CONSTEXPR(sizeof(T) == 4)
             {
-                return _mm512_mask_sub_epi32(self, mask.data, self, _mm512_set1_epi32(1));
+                return _mm512_mask_sub_epi32(self, mask, self, _mm512_set1_epi32(1));
```
this sneaked in. I'll revert. My OCD clicked in...

serge-sans-paille commented:
Merged as 239fbbd96dda5bb48a6c9a205e93836a724de351
Thanks a lot @DiamonDinoia for the hard work. It's now on me to implement part of it for other architectures.

DiamonDinoia commented on include/xsimd/arch/xsimd_avx2.hpp:
```diff
@@ -116,6 +117,137 @@ namespace xsimd
+        XSIMD_INLINE batch<int32_t, A> load_masked(int32_t const* mem, batch_bool_constant<int32_t, A, Values...> mask, convert<int32_t>, Mode, requires_arch<avx2>) noexcept
+        {
+            return load_masked<A>(mem, mask, convert<int32_t> {}, Mode {}, avx2 {});
+        }
```
I might have meant calling uint -> int32 there.
